Consume as flow

elizarov opened the pull request.

qwwdfsad commented on kotlinx-coroutines-core/common/test/flow/channels/ChannelBuildersFlowTest.kt:
```diff
@@ -9,6 +9,51 @@ import kotlinx.coroutines.channels.*
+        val flow = channel.consumeAsFlow()
+        assertFailsWith<TestException> { flow.sum() }
+        assertFailsWith<IllegalStateException> { flow.collect() }
+        finish(2)
```
not sure why these `expect(1)` + `finish(2)` are needed

elizarov replied:
Not really needed. Just for consistency.

qwwdfsad commented on kotlinx-coroutines-core/common/src/channels/Channel.kt:
```diff
@@ -242,6 +243,19 @@ public interface ReceiveChannel<out E> {
+     * @suppress **This an internal API and should not be used from general code.**
+     */
+    @InternalCoroutinesApi
+    public suspend fun consumeEachTo(collector: FlowCollector<E>)
```
I wonder why we need such intrusive method for `@FlowPreview` API.
We can write `cosumeAsFlow` on top of an iterator with nulling-out the slot to workaround #1333.
I realize that such an approach is a bit slower, but at least it does not require to add a new internal method to already overloaded channels

elizarov replied:
For a fix to #1333 it can be indeed just an extension. However, converting to/from channels seems to be important part of a flow API surface and writing it in such a way offers better performance as it avoids iterator altogether. Also, the whole state machine and spilling around iterator-using code is much more complicated and the resulting code would be even more kludgy. 

However, there are other way to do the fix. An alternative but also quite performant way would be to introduce `ReceiveOrClosed` (even as experimental/iternal API that would be good here), and then `consumeEachTo` can be written as extension in the flow package. Can we resurrect PR #762 ?

zach-klippenstein replied:
I would selfishly love this alternative, I've got one or two other use cases for `ReceiveOrClosed` that I'm working around awkwardly right now.

elizarov replied:
Pulled in PR #762 into here and implemented it in a less intrusive way.

elizarov commented on kotlinx-coroutines-core/common/src/channels/AbstractChannel.kt:
```diff
@@ -663,6 +640,57 @@ internal abstract class AbstractChannel<E> : AbstractSendChannel<E>(), Channel<E
+                //     L$1 <- collector
+                //     L$2 <- cause
+                //     L$3 <- this$run (actually equal to this)
+                val result = run { receiveInternal() }
```
Tried it, but it does not work. It does not spill the old result before call `receveInternal` even if it a `var` (it knows it is going to be overwritten anyway on resume), so setting it to `null` offers no benefit.

zach-klippenstein commented on kotlinx-coroutines-core/common/src/channels/Channel.kt:
```diff
@@ -271,6 +323,82 @@ public interface ReceiveChannel<out E> {
+        get() = if (isClosed) null else holder as T
+
+    @Suppress("UNCHECKED_CAST")
+    public val valueOrThrow: T
```
How is this actually different from `value` in the closed case? It's possible to figure out that the exceptions are different from reading the code, but would be helpful to have that document in the kdoc. Even reading the code, I'm not sure when I'd use this property instead of just `value`.

elizarov replied:
This is not critical to resolve right now (since it is internal). I'm leaving todo there.

zach-klippenstein commented on kotlinx-coroutines-core/common/src/channels/Channel.kt:
```diff
@@ -271,6 +323,82 @@ public interface ReceiveChannel<out E> {
+     * Returns received value if this element represents received value or `null` otherwise.
+     */
+    @Suppress("UNCHECKED_CAST")
+    public val valueOrNull: T?
```
Nit: If this were an extension function on `<T: Any> ReceiveResult<T>` (i.e. when `T` is non-nullable), it would prevent the mistake of conflating closed with null, which this class is meant to solve in the first place.  Is that too restrictive? Given that `value` already does the safe thing anyway, I guess this verbose name makes it really clear you're acknowledging that.

elizarov replied:
Leaving todo.

zach-klippenstein commented on kotlinx-coroutines-core/common/src/channels/Channel.kt:
```diff
@@ -271,6 +310,82 @@ public interface ReceiveChannel<out E> {
+
+    internal companion object {
+        @Suppress("NOTHING_TO_INLINE")
+        internal inline fun <E> value(value: E): ReceiveResult<E> =
```
Would you consider making these public once this API is not longer `@InternalCoroutinesApi`? Could be useful to be able to create these for unit tests.

elizarov replied:
Don't see any problem with, but let's track that separately as a future enchancement. We cannot make it all public anyway now, due to instability of inline classes ABI.

qwwdfsad commented on kotlinx-coroutines-core/common/src/channels/Channel.kt:
```diff
@@ -271,6 +322,101 @@ public interface ReceiveChannel<out E> {
+     *
+     * :todo: Decide if it is needed how it shall be named with relation to [valueOrThrow]:
+     *
+     * So we have the following methods on ValueOrClosed: `value`, `valueOrNull`, `valueOrThrow`.
```
I am ok with postponing these TODOs for now while this API is unstable, so feel free to resolve my comments regarding it

qwwdfsad commented on kotlinx-coroutines-core/common/src/channels/Channel.kt:
```diff
@@ -271,6 +322,101 @@ public interface ReceiveChannel<out E> {
+     *  * `receive` which corresponds to `receiveOrClosed().valueOrThrow`... huh?
+     *  * `receiveOrNull` which corresponds to `receiveOrClosed().valueOrNull`
+     *  * `receiveOrClosed`
+     * For the sake of consider dropping this version of `value` and rename [valueOrThrow] to simply `value`.
```
I'd rather avoid dropping this version. `valueOrThrow` **rethrows** the `cause` adding no meaningful information about the callsite, so if one is sure that `ValueOrClosed` is always `value`, this very property should be used. Otherwise, it could be very hard to locate the source of the exception

elizarov replied:
Added the above comment to todos.

qwwdfsad commented on kotlinx-coroutines-core/common/src/channels/Channel.kt:
```diff
@@ -271,6 +322,101 @@ public interface ReceiveChannel<out E> {
+    /**
+     * Returns received value if this element represents received value or `null` otherwise.
+     * :todo: Decide if it shall be made into extension that is available only for non-null T.
+     * (note: it might become inconsistent with kotlin.Result)
```
I'd rather make it inconsistent with `Result`, but safe.
`Result.getOrNull` was added as member deliberately?

qwwdfsad commented on kotlinx-coroutines-core/common/src/channels/Channel.kt:
```diff
@@ -271,6 +322,101 @@ public interface ReceiveChannel<out E> {
+     */
+    @Suppress("UNCHECKED_CAST")
+    public val value: T
+        get() = if (isClosed) throw IllegalStateException() else holder as T
```
`if (isClosed) error(...)`

qwwdfsad commented on kotlinx-coroutines-core/common/src/flow/Channels.kt:
```diff
@@ -7,14 +7,81 @@
+ * 1) Flow consumer is cancelled when the original channel is cancelled.
+ * 2) Flow consumer completes normally when the original channel completes (~is closed) normally.
+ * 3) If the flow consumer fails with an exception, channel is cancelled.
+ *
```
extra indent

elizarov replied:
Where?

qwwdfsad replied:
line 74

qwwdfsad commented on kotlinx-coroutines-core/common/src/flow/Channels.kt:
```diff
@@ -7,14 +7,81 @@
+ *
+ */
+@FlowPreview
+public fun <T> ReceiveChannel<T>.consumeAsFlow(): Flow<T> = object : Flow<T> {
```
[optional] I find out that reading flow code is really unpleasant.
E.g. this method adds a frame named `kotlinx.coroutines.flow.FlowKt__ChannelsKt$consumeAsFlow$1.collect(Channels.kt:81)`. Not the best user experience :) 

While we can avoid such frames for inline methods, we can at least do a best-effort on non-inline ones

elizarov replied:
How can it be fixed? Move the code to a named class?

romansl commented on kotlinx-coroutines-core/common/src/channels/Channel.kt:
```diff
@@ -91,7 +93,8 @@ public interface SendChannel<in E> {
      *
-     * A channel that was closed without a [cause] throws [ClosedSendChannelException] on attempts to send or receive.
+     * A channel that was closed without a [cause] throws [ClosedSendChannelException] on attempts to send or
+     * and [ClosedReceiveChannelException] on attempts to receive.
```
Typo: odd `or` before `and`.

elizarov replied:
Thanks!

elizarov commented:
I've squashed and rearranged all the old commit into two commits, fixing all the review issues:

1. "Channel.receiveOrNull becomes extension, internal receiveOrClosed added" combines all the improvement to channels machinery and introduces internal `receiveOrClosed`.
2. "Introduce ReceiveChannel.consumeAsFlow and FlowCollector.emitAll(chan)" implements `consumeAsFlow` using those channel improvements.

I've added a 3rd (new) commit that makes `consumeAsFlow` fuseable with all the channel-using flow operators so it avoids creation of an extra channel when possible and unless explicitly requested. Please, take a look.
